geo/src/algorithm/outlier_detection.rs

```diff
@@ -0,0 +1,302 @@
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "Point index: {}, LOF Score: {:?}",
+            self.index, self.score
+        )
+    }
+}
+
+fn lof<T>(points: &[Point<T>], kneighbours: usize) -> Vec<LofScore<T>>
+where
+    T: GeoFloat + Sum,
+{
+    if points.len() <= kneighbours {
+        // no point in trying to run the algorithm in this case
+        return points
+            .iter()
+            .enumerate()
+            .map(|(idx, _)| LofScore::new(idx, T::zero()))
+            .collect();
+    }
+    // we need to keep track of original indices, so create a custom type
+    let geoms: Vec<GeomWithData<_, usize>> = points
+        .iter()
+        .enumerate()
+        .map(|(idx, point)| GeomWithData::new(*point, idx))
+        .collect();
+    let tree = RTree::bulk_load(geoms);
+    // get (GWD, Distance2) – and thus also kth-nearest neighbour distance as a vec of tuples for each point.
+    let knn_dists = points
+        .iter()
+        .map(|point| {
+            tree.nearest_neighbor_iter_with_distance_2(point)
+                .take(kneighbours)
+                .collect()
+        })
+        .collect::<Vec<Vec<_>>>();
+    // calculate LRD (local reachability density) of each point
+    // LRD is the estimated distance at which a point can be found by its neighbours:
+    // count(neighbour_set) / sum(max(point.kTh_dist, point.dist2(other point)) for all points in neighbour_set)
+    // we call this sum-of–max-distances reachDistance
+    let local_reachability_densities: Vec<T> = knn_dists
+        .iter()
+        .map(|neighbours| {
+            // for each point's neighbour set, calculate kth distance
+            let kth_dist = neighbours
+                .iter()
+                .map(|(_, distance)| distance)
+                .last()
+                .unwrap();
+            T::from(neighbours.len()).unwrap()
+                / neighbours
+                    .iter()
+                    // sum the max between neighbour distance and kth distance for the neighbour set
+                    .map(|(_, distance)| distance.max(*kth_dist))
+                    .sum()
+        })
+        .collect();
+    // LOF of a point p is the sum of the LRD of all the points
+    // in the set kNearestSet(p) * the sum of the reachDistance of all the points of the same set,
+    // to the point p, all divided by the number of items in p's kNN set, squared.
+    knn_dists
+        .iter()
+        .enumerate()
+        .map(|(idx, neighbours)| {
+            // for each point's neighbour set, calculate kth distance
+            let kth_dist = neighbours
+                .iter()
+                .map(|(_, distance)| distance)
+                .last()
+                .unwrap();
+            // sum neighbour set LRD scores
+            let lrd_scores: T = neighbours
+                .iter()
+                .map(|(neighbour, _)| local_reachability_densities[neighbour.data])
```

Comment:
I suspect there's a way to calculate LRDs for a subset ( the point's neighbours) as we go, but I can't get it right so we're stuck with `local_reachability_densities` and a redundant `kth_dist` calculation for now. I think it would ultimately end up being two passes over the data anyway, but it might use less space.